services-turf/src/commonMain/kotlin/org/maplibre/geojson/util/Measurement.kt

```diff
@@ -0,0 +1,551 @@
+ *
+ * @param line1StartX X coordinate of line 1 start.
+ * @param line1StartY Y coordinate of line 1 start.
+ * @param line1EndX X coordinate of line 1 end.
+ * @param line1EndY Y coordinate of line 1 end.
+ * @param line2StartX X coordinate of line 2 start.
+ * @param line2StartY Y coordinate of line 2 start.
+ * @param line2EndX X coordinate of line 2 end.
+ * @param line2EndY Y coordinate of line 2 end.
+ * @return [LineIntersects] if intersection exists, null otherwise.
+ */
+private fun lineIntersects(
+    line1StartX: Double,
+    line1StartY: Double,
+    line1EndX: Double,
+    line1EndY: Double,
+    line2StartX: Double,
+    line2StartY: Double,
+    line2EndX: Double,
+    line2EndY: Double
+): LineIntersects? {
+    // If the lines intersect, the result contains the x and y of the intersection
+    // (treating the lines as infinite) and booleans for whether line segment 1 or line
+    // segment 2 contain the point
+    var result = LineIntersects(
+        onLine1 = false,
+        onLine2 = false
+    )
+
+    val denominator = (((line2EndY - line2StartY) * (line1EndX - line1StartX))
+            - ((line2EndX - line2StartX) * (line1EndY - line1StartY)))
+    if (denominator == 0.0) {
+        return if (result.horizontalIntersection != null && result.verticalIntersection != null) {
+            result
+        } else {
+            null
+        }
+    }
+    var varA = line1StartY - line2StartY
+    var varB = line1StartX - line2StartX
+    val numerator1 = ((line2EndX - line2StartX) * varA) - ((line2EndY - line2StartY) * varB)
+    val numerator2 = ((line1EndX - line1StartX) * varA) - ((line1EndY - line1StartY) * varB)
+    varA = numerator1 / denominator
+    varB = numerator2 / denominator
+
+    // if we cast these lines infinitely in both directions, they intersect here:
+    result = result.copy(
+        horizontalIntersection = line1StartX + (varA * (line1EndX - line1StartX)),
+        verticalIntersection = line1StartY + (varA * (line1EndY - line1StartY))
+    )
+
+    // if line1 is a segment and line2 is infinite, they intersect if:
+    if (varA > 0 && varA < 1) {
+        result = result.copy(onLine1 = true)
+    }
+    // if line2 is a segment and line1 is infinite, they intersect if:
+    if (varB > 0 && varB < 1) {
+        result = result.copy(onLine2 = true)
+    }
+    // if line1 and line2 are segments, they intersect if both of the above are true
+    return if (result.onLine1 && result.onLine2) {
+        result
+    } else {
+        null
+    }
+}
+
+/**
+ * Data class representing the result of a line intersection.
+ *
+ * @property horizontalIntersection The X coordinate of the intersection.
+ * @property verticalIntersection The Y coordinate of the intersection.
+ * @property onLine1 True if the intersection is on the first line segment.
+ * @property onLine2 True if the intersection is on the second line segment.
+ */
+data class LineIntersects(
```

Comment:
```suggestion
private data class LineIntersects(
```
doesn't need to be exposed as long as it is only used internally